### ISSUE TITLE
[Bug]: UI partially not working when behind Authelia

### ISSUE THREAD

**Issue by MMauro94** [tubearchivist/tubearchivist#846]
### I've read the documentation

- [X] I'm running the latest version of Tube Archivist and have read the [release notes](https://github.com/tubearchivist/tubearchivist/releases/latest).
- [X] I have read the [how to open an issue](https://github.com/tubearchivist/tubearchivist/blob/master/CONTRIBUTING.md#how-to-open-an-issue) guide, particularly the [bug report](https://github.com/tubearchivist/tubearchivist/blob/master/CONTRIBUTING.md#bug-report) section.

### Operating System

Linux Debian

## Describe the bug

The UI is partially not working when adding Authelia in front of Tube Archivist. More specifically, it looks like all JavaScipt requests against `/api` get back with an HTTP return code of `302` and with an empty `Location` header. The browser tries a few times and then gives up due to a redirect loop error.

I've investigated a little bit myself, and found out that the `Authorization` header might me the culprit. By copying the `curl` request from the network request in the browser, I was able to reproduce the issue in a short shell command.

Does not work (removed some useless headers):
```bash
curl -I 'https://<my-instance-url>/api/video/' \
  -H 'Content-Type: application/json' \
  -H 'Authorization: Token' \
  -H 'Cookie: authelia_session=<REDACTED>; csrftoken=<REDACTED>; sessionid=<REDACTED>'
HTTP/2 302
server: openresty
date: Sat, 14 Dec 2024 14:42:15 GMT
content-type: text/html
content-length: 142
www-authenticate: Basic realm="Authorization Required"
location:
```

I noticed that the empty token in the `Authorization` header, so I tried to remove it... and voilà:
```bash
curl -I 'https://<my-instance-url>/api/video/' \
  -H 'Content-Type: application/json' \
  -H 'Cookie: authelia_session=<REDACTED>; csrftoken=<REDACTED>; sessionid=<REDACTED>'
HTTP/2 200
server: openresty
date: Sat, 14 Dec 2024 14:46:51 GMT
content-type: application/json
content-length: 40493
vary: Accept, origin, Cookie
allow: GET, HEAD, OPTIONS
x-frame-options: DENY
x-content-type-options: nosniff
referrer-policy: same-origin
cross-origin-opener-policy: same-origin
```

I found the code setting that header [here](https://github.com/tubearchivist/tubearchivist/blob/744c6c94f1c5450f4f043478601a8dccb700d91a/tubearchivist/static/script.js#L1061); however, that `sessionid` cookie is set as [HttpOnly](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie#httponly) (at least that's what I see in my browser). So that is why `getCookie('sessionid')` returns an empty string and the header is sent without a token.

Quoting from [this docs](https://docs.tubearchivist.com/api/introduction/#authentication):
> session based authentication is enabled too: When you are logged into your TubeArchivist instance, you'll have access to the api in the browser for testing

So I'm not understanding why that code is there at all, as that couldn't possibly work given that that cookie is set as HttpOnly. Also I couldn't find where that `sessionid` token is set in code, is it a Django thing? Or maybe it comes from a library and that's why I couldn't find it.

Opening `https://<my-instance-url>/api/video/` in my browser shows the Djang REST UI thingy and works normally.

## Steps To Reproduce
1. Put TubeArchivist behind Authelia. I am personally using [NginxProxyManager](https://nginxproxymanager.com/) as a reverse proxy, with the following custom config:
```
include /snippets/authelia/authelia-location.conf;

location / {
    include /snippets/authelia/proxy.conf;
    include /snippets/authelia/authelia-authrequest.conf;
    proxy_pass $forward_scheme://$server:$port;
}
```
The snippets are verbatim what is found in the Authelia website [here](https://www.authelia.com/integration/proxies/nginx/#supporting-configuration-snippets).

2. Open tube archivist and try to reproduce a video. See in the network tab that the request to `/api/video/<video-id>` doesn't work.

Let me know if you want more info from me. There's nothing in the logs as far as I can see.

## Expected behavior

Works :)

### Relevant log output

```shell
machine: x86_64
clock source: unix
detected number of CPU cores: 16
current working directory: /app
writing pidfile to /tmp/project-master.pid
detected binary path: /root/.local/bin/uwsgi
!!! no internal routing support, rebuild with pcre support !!!
uWSGI running as root, you can use --uid/--gid/--chroot options
*** WARNING: you are running uWSGI as root !!! (use the --uid flag) *** 
your memory page size is 4096 bytes
detected max file descriptor number: 1048576
lock engine: pthread robust mutexes
thunder lock: disabled (you can enable it with --thunder-lock)
uwsgi socket 0 bound to TCP address :8080 fd 3
uWSGI running as root, you can use --uid/--gid/--chroot options
*** WARNING: you are running uWSGI as root !!! (use the --uid flag) *** 
Python version: 3.11.8 (main, Mar 12 2024, 11:52:02) [GCC 12.2.0]
Python main interpreter initialized at 0x7fea01d51218
uWSGI running as root, you can use --uid/--gid/--chroot options
*** WARNING: you are running uWSGI as root !!! (use the --uid flag) *** 
python threads support enabled
your server socket listen backlog is limited to 100 connections
your mercy for graceful operations on workers is 60 seconds
mapped 154032 bytes (150 KB) for 1 cores
*** Operational MODE: single process ***
celery beat v5.4.0 (opalescent) is starting.
WSGI app 0 (mountpoint='') ready in 0 seconds on interpreter 0x7fea01d51218 pid: 43 (default app)
uWSGI running as root, you can use --uid/--gid/--chroot options
*** WARNING: you are running uWSGI as root !!! (use the --uid flag) *** 
*** uWSGI is running in multiple interpreter mode ***
spawned uWSGI master process (pid: 43)
spawned uWSGI worker 1 (pid: 61, cores: 1)
__    -    ... __   -        _
LocalTime -> 2024-12-14 14:32:56
Configuration ->
    . broker -> redis://tubearchivist-redis:6379//
    . loader -> celery.loaders.app.AppLoader
    . scheduler -> django_celery_beat.schedulers.DatabaseScheduler
    . logfile -> [stderr]@%INFO
    . maxinterval -> 5.00 seconds (5s)
[2024-12-14 14:32:56,011: INFO/MainProcess] beat: Starting...
/root/.local/lib/python3.11/site-packages/celery/platforms.py:829: SecurityWarning: You're running the worker with superuser privileges: this is
absolutely not recommended!
Please specify a different user using the --uid option.
User information: uid=0 euid=0 gid=0 egid=0
  warnings.warn(SecurityWarning(ROOT_DISCOURAGED.format(
 
 -------------- celery@fbe39c3f4ff0 v5.4.0 (opalescent)
--- ***** ----- 
-- ******* ---- Linux-6.1.0-26-amd64-x86_64-with-glibc2.36 2024-12-14 14:32:56
- *** --- * --- 
- ** ---------- [config]
- ** ---------- .> app:         tasks:0x7ffbd8fe0110
- ** ---------- .> transport:   redis://tubearchivist-redis:6379//
- ** ---------- .> results:     redis://tubearchivist-redis:6379/
- *** --- * --- .> concurrency: 16 (prefork)
-- ******* ---- .> task events: OFF (enable -E to monitor tasks in this worker)
--- ***** ----- 
 -------------- [queues]
                .> celery           exchange=celery(direct) key=celery
                
[tasks]
  . check_reindex
  . download_pending
  . extract_download
  . index_playlists
  . manual_import
  . rescan_filesystem
  . restore_backup
  . resync_thumbs
  . run_backup
  . subscribe_to
  . thumbnail_check
  . update_subscribed
  . version_check
[2024-12-14 14:32:56,580: WARNING/MainProcess] /root/.local/lib/python3.11/site-packages/celery/worker/consumer/consumer.py:508: CPendingDeprecationWarning: The broker_connection_retry configuration setting will no longer determine
whether broker connection retries are made during startup in Celery 6.0 and above.
If you wish to retain the existing behavior for retrying connections on startup,
you should set broker_connection_retry_on_startup to True.
  warnings.warn(
[2024-12-14 14:32:56,582: INFO/MainProcess] Connected to redis://tubearchivist-redis:6379//
[2024-12-14 14:32:56,583: WARNING/MainProcess] /root/.local/lib/python3.11/site-packages/celery/worker/consumer/consumer.py:508: CPendingDeprecationWarning: The broker_connection_retry configuration setting will no longer determine
whether broker connection retries are made during startup in Celery 6.0 and above.
If you wish to retain the existing behavior for retrying connections on startup,
you should set broker_connection_retry_on_startup to True.
  warnings.warn(
[2024-12-14 14:32:56,584: INFO/MainProcess] mingle: searching for neighbors
[2024-12-14 14:32:57,588: INFO/MainProcess] mingle: all alone
[2024-12-14 14:32:57,597: INFO/MainProcess] celery@fbe39c3f4ff0 ready.
Sat Dec 14 14:46:37 2024 - SIGPIPE: writing to a closed pipe/socket/fd (probably the client disconnected) on request /static/rest_framework/css/bootstrap.min.f17d4516b026.css (ip 172.18.0.43) !!!
Sat Dec 14 14:46:37 2024 - uwsgi_response_sendfile_do(): Broken pipe [core/writer.c line 655] during GET /static/rest_framework/css/bootstrap.min.f17d4516b026.css (172.18.0.43)
OSError: write error
```

**Comment by bbilly1** [comment]
Who returns the 302? Authelia? Nginx? Uwsgi? 

Requests are authenticated in TA if you pass a valid session ID or a valid Token. Even if you pass an invalid Token authentication header but a valid seesion ID, this should still work, at least from the perspective of TA.

I would assume, that authelia somehow does not forward the session ID correctly. 

But ultimatly, I don't have any experience with authelia. If somebody has more experience with that, please share.

**Comment by MMauro94** [comment]
Ack, thanks for the info. I will dive a bit deeper in what's happening and report back.

**Comment by MMauro94** [comment]
So I've dug a bit deeper into this.

From what I could see:
- the `302` is returned by Authelia (more specifically by the [`authelia-authrequest.conf` snippet](https://www.authelia.com/integration/proxies/nginx/#authelia-authrequestconf) whenever Authelia returns `401`)
- it does that because the `Authentication` header is malformed (see log below)
- the `Location` header is empty because Authelia itself doesn't return a `Location` header in this case. I believe this is by design, as the URL parameter in [this Authelia function](https://github.com/authelia/authelia/blob/master/internal/handlers/handler_authz_authn.go#L286) is explicitly ignored. So the `$redirection_url` in the script will be empty an empty `Location` is sent back.
  - This is obviously completely out of scope for TubeArchivist, I will open a separate issue in the Authelia repo

On this request:

```bash
curl -I 'https://<my-instance-url>/api/video/' \
  -H 'Content-Type: application/json' \
  -H 'Authorization: Token' \
  -H 'Cookie: authelia_session=<REDACTED>; csrftoken=<REDACTED>; sessionid=<REDACTED>'
HTTP/2 302
server: openresty
date: Sat, 14 Dec 2024 14:42:15 GMT
content-type: text/html
content-length: 142
www-authenticate: Basic realm="Authorization Required"
location:
```

I get this output from Authelia:

```
time="2024-12-27T15:09:18Z" level=trace msg="Request hit" caller="github.com/authelia/authelia/v4/internal/middlewares/log_request.go:12 handleRouter.LogRequest.func40" method=GET path=/api/authz/auth-request remote_ip=<REDACTED>
time="2024-12-27T15:09:18Z" level=debug msg="Check authorization of subject username= groups= ip=<REDACTED> and object https://<REDACTED>/api/video/ (method HEAD)." caller="github.com/authelia/authelia/v4/internal/authorization/authorizer.go:52 (*Authorizer).GetRequiredLevel"
time="2024-12-27T15:09:18Z" level=trace msg="ACL MISS Position 1 for subject username= groups= ip=<REDACTED> and object https://<REDACTED>/api/video/ (method HEAD, policy two_factor)" caller="github.com/authelia/authelia/v4/internal/authorization/authorizer.go:62 (*Authorizer).GetRequiredLevel"
time="2024-12-27T15:09:18Z" level=trace msg="ACL HIT Position 2 for subject username= groups= ip=<REDACTED> and object https://<REDACTED>/api/video/ (method HEAD, policy two_factor)" caller="github.com/authelia/authelia/v4/internal/authorization/authorizer.go:57 (*Authorizer).GetRequiredLevel"
time="2024-12-27T15:09:18Z" level=debug msg="Error occurred while attempting to authenticate a request but the matched rule was a bypass rule" caller="github.com/authelia/authelia/v4/internal/handlers/handler_authz.go:91 (*Authz).Handler" error="failed to parse content of Authorization header: invalid scheme: the scheme is missing" method=GET path=/api/authz/auth-request remote_ip=<REDACTED>
time="2024-12-27T15:09:18Z" level=debug msg="Responding 401 WWW-Authenticate" caller="github.com/authelia/authelia/v4/internal/handlers/handler_authz_authn.go:287 (*HeaderAuthnStrategy).HandleUnauthorized" method=GET path=/api/authz/auth-request remote_ip=<REDACTED>
time="2024-12-27T15:09:18Z" level=trace msg="Replied (status=401)" caller="github.com/authelia/authelia/v4/internal/middlewares/log_request.go:16 handleRouter.LogRequest.func40" method=GET path=/api/authz/auth-request remote_ip=<REDACTED>
```

So apparently this boils down to Authelia blowing up whenever it's passed an invalid `Authorization` header. Since I'm behind NGINX, it is configured to hit the `/api/authz/auth-request` Authelia endpoint (see doc https://www.authelia.com/reference/guides/proxy-authorization/#default-endpoints).

I've confirmed this by calling the Authelia endpoint directly:
```bash
curl -I -X GET "http://authelia:9091/api/authz/auth-request" \
  -H 'X-Original-Method: GET' \
  -H 'X-Original-URL: https://<REDACTED>' \
  -H 'X-Forwarded-For: 127.0.0.1' \
  -H 'Cookie: authelia_session=<REDACTED>; csrftoken=<REDACTED>; sessionid=<REDACTED>' \
  -H 'Authorization: Token'
HTTP/1.1 401 Unauthorized
Date: Fri, 27 Dec 2024 15:27:06 GMT
Content-Type: text/plain; charset=utf-8
Content-Length: 16
Www-Authenticate: Basic realm="Authorization Required"
```

Again, removing the `Authorization` header works, or even just "fixing" it with trash data:
```bash
curl -I -X GET "http://authelia:9091/api/authz/auth-request" \
  -H 'X-Original-Method: GET' \
  -H 'X-Original-URL: https://<REDACTED>' \
  -H 'X-Forwarded-For: 127.0.0.1' \
  -H 'Cookie: authelia_session=<REDACTED>; csrftoken=<REDACTED>; sessionid=<REDACTED>' \
  -H 'Authorization: Bearer abcd'
HTTP/1.1 200 OK
Date: Fri, 27 Dec 2024 15:29:45 GMT
Content-Type: text/plain; charset=utf-8
Content-Length: 6
Remote-User: <REDACTED>
Remote-Groups: <REDACTED>,<REDACTED>
Remote-Name: <REDACTED>
Remote-Email: <REDACTED>
```

Notice that part of the fix was change `Token` to `Bearer`, otherwise I would still get an error from Authelia:
```
time="2024-12-27T15:44:26Z" level=debug msg="Error occurred while attempting to authenticate a request but the matched rule was a bypass rule" caller="github.com/authelia/authelia/v4/internal/handlers/handler_authz.go:91 (*Authz).Handler" error="failed to parse content of Authorization header: invalid scheme: scheme with name 'token' is unknown" method=GET path=/api/authz/auth-request remote_ip=127.0.0.1
```

What's still unclear to me is why it's trying to parse `Authorization` rather than `Proxy-Authorization`, which according to [this doc](https://www.authelia.com/reference/guides/proxy-authorization/#headerauthrequestproxyauthorization) is what it's supposed doing, but that's again beyond the scope of this issue.


----

While this is more of a bug in Authelia (I see no reason why it a malformed header should completely fail the request instead of just going to the next auth strategy), I still find it odd that TA is sending a malformed `Authorization` header (invalid scheme + empty token due to the cookie being set as HttpOnly).

In this state, it looks like that [this line](https://github.com/tubearchivist/tubearchivist/blob/744c6c94f1c5450f4f043478601a8dccb700d91a/tubearchivist/static/script.js#L1061) and line 1064 are completely useless (correct me if I'm wrong).

Would you be open to remove them? This would at least for now be a workaround for putting TA behind Authelia, and just a matter of cleaning up the not needed code.

**Comment by MMauro94** [comment]
So as per Authelia issue above, the TL;DR is:
- By design, Authelia stops and returns an error whenever it encounters a serious issue with one authentication strategy (e.g. malformed header), and doesn't try other strategies
  - In this case, it returns a `WWW-Authenticate` header and _no_ `Location` header, as per spec
- The empty `Location` header is returned due to a limitation in NGINX's `auth_request` module (can't do `ifs` on the auth response)

I've fixed my specific issue by overriding the authorization endpoint and removing the header authorization (which I don't need in any of my services), as suggested by the Authelia team:
```yaml
server:
  endpoints:
    authz:
      auth-request:
        implementation: 'AuthRequest'
        authn_strategies:
          - name: 'CookieSession'
```

While this works for my use-case, I would still argue that [this line](https://github.com/tubearchivist/tubearchivist/blob/744c6c94f1c5450f4f043478601a8dccb700d91a/tubearchivist/static/script.js#L1061) is unnecessary, as the `sessionid` token is set as httponly, and anyone who tries to add Authelia in front of TubeArchivist is going to find this issue.

**Comment by bbilly1** [comment]
Thanks for looking into this. I indeed do not know why the line is there as you pointed out as only the `X-CSRFToken` header is really needed. That's probably some leftovers of some experiment that got unnoticed as it _usually_ doesn't have any effect. 

As you saw, or main focus is on the react migrations: https://github.com/tubearchivist/tubearchivist/issues/763. As we are making good progress there, it's unlikely that this will be addressed here in the mean time. But I'm glad I found a workaround for your case.

That will use regular session based authentication and should not set any strange headers like that.

**Comment by MMauro94** [comment]
Thanks for this awesome project :) Yeah that makes sense. I'm happy with the workaround for now. I'm glad that the new frontend is going well, can't wait to try that out!

Up to you if you want to keep this issue around or close it as this is going to get addressed anyway eventually.

Thanks and happy new year :)

**Comment by KerbinCenturion** [comment]
This problem is exclusive to Authelia and all programs that had similar issues were gone once I switched to Authentik. While it is a bit more of a faf to set up once it is done it will save you a lot of headaches.

**Comment by bbilly1** [comment]
OK, this should be fixed in v0.5.0, please update.

**Comment by MMauro94** [comment]
Can confirm this issue is resovled in v0.5.1. I've removed the [snippet from above](https://github.com/tubearchivist/tubearchivist/issues/846#issuecomment-2575985501) in my Authelia config and everything still works.

Thanks!
